File: experiment3.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<string> possible_combinations = {
    "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z",
};

vector<string> &getAllAlphabetCombinations(int length) {
  cout << "making list of length " << length << endl;
  vector<string> *result = new vector<string>();
  while (length > 0) {
    if (result->size() == 0) {
      for (int i = 0; i < 26; i++) {
        result->push_back(possible_combinations[i]);
      }
    } else {
      vector<string> *newResult = new vector<string>();
      for (int i = 0; i < result->size(); i++) {
        for (int alphabetIdx = 0; alphabetIdx < 26; alphabetIdx++) {
          newResult->push_back(result->at(i) + possible_combinations[alphabetIdx]);
        }
      }
      result = newResult;
    }
    length--;
  }
  return *result;
}

bool hasAllAlphabetCombinations(string &inputVal) {
  for (int i = 0; i < 26; i++) {
    for (int j = 0; j < 26; j++) {
      string s = possible_combinations[i] + possible_combinations[j];
      if (inputVal.find(s) == string::npos) {
        return false;
      }
    }
  }
  return true;
}

int main() {
  int search_string_lower_bound = 2 * 26;       // 26 letters, 2 times each
  int search_string_upper_bound = 2 * 26 * 26;  // 26 letters, 2 times each, 26 combinations

  for (int i = search_string_lower_bound; i < search_string_upper_bound; i++) {
    vector<string> candidates = getAllAlphabetCombinations(i);
    for (int j = 0; j < candidates.size(); j++) {
      cout << candidates[j] << endl;
      if (hasAllAlphabetCombinations(candidates[j])) {
        cout << "Found it: " << candidates[j] << endl;
        return 0;
      }
    }
  }
  return 0;
}

File: experiment3.py
possible_combinations = [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
    "h",
    "i",
    "j",
    "k",
    "l",
    "m",
    "n",
    "o",
    "p",
    "q",
    "r",
    "s",
    "t",
    "u",
    "v",
    "w",
    "x",
    "y",
    "z",
]


def get_all_alphabet_combinations_list(length):
    print(f"making list of length {length}")
    result = []
    while length > 0:
        if len(result) == 0:
            result = possible_combinations
        else:
            new_result = []
            for i, v in enumerate(result):
                for i2, v2 in enumerate(possible_combinations):
                    new_result.append(v + v2)
            result = new_result
        length -= 1
    return result


def has_all_alphabet_combinations(string):
    for _, v1 in enumerate(possible_combinations):
        for _, v2 in enumerate(possible_combinations):
            if v1 + v2 not in string:
                return False
    return True


search_string_lower_bound = 2 * 26  # 26 letters, 2 times each
search_string_upper_bound = 2 * 26 * 26  # 26 letters, 2 times each, 26 combinations

for i in range(search_string_lower_bound, search_string_upper_bound):
    candidates = get_all_alphabet_combinations_list(i)
    for _, v in enumerate(candidates):
        print(v)
        if has_all_alphabet_combinations(v):
            print("Found it: ", v)
            exit(0)
